iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_fifo.inl

```diff
@@ -63,27 +105,50 @@ inline constexpr uint64_t SpscFifo<ValueType, Capacity>::capacity() noexcept
 {
-    return m_read_pos.load(std::memory_order_relaxed) == m_write_pos.load(std::memory_order_relaxed);
+    auto [readPos, writePos] = get_read_write_positions();
+    return readPos == writePos;
 }
 
 template <class ValueType, uint64_t Capacity>
 inline optional<ValueType> SpscFifo<ValueType, Capacity>::pop() noexcept
 {
-    auto currentReadPos = m_read_pos.load(std::memory_order_acquire);
-    bool isEmpty = (currentReadPos ==
-                    // we are not allowed to use the empty method since we have to sync with
-                    // the producer pop - this is done here
-                    m_write_pos.load(std::memory_order_acquire));
-    if (isEmpty)
+    // Memory order relaxed is enough since:
+    // - there is no concurrent access to this method
+    // - the load statement cannot be reordered with the is_empty check otherwise it would not
+    // compile
+    auto current_readPos = m_readPos.load(std::memory_order_relaxed);
+
+    bool is_empty = (current_readPos ==
+                     // SYNC POINT READ: m_data
+                     // See explanation of the corresponding sync point.
+                     // As a consequence, we are not allowed to use the empty method
+                     // since we have to sync with m_writePos in the push method
+                     m_writePos.load(std::memory_order_acquire));
+    if (is_empty)
     {
         return nullopt_t();
     }
-    ValueType out = m_data[currentReadPos % Capacity];
 
-    // m_read_pos must be increased after reading the pop'ed value otherwise
-    // it is possible that the pop'ed value is overwritten by push while it is read.
-    // Implementing a single consumer fifo here allows us to use store.
-    m_read_pos.store(currentReadPos + 1, std::memory_order_release);
+    ValueType out = m_data[current_readPos % Capacity];
+
+    // We need to make sure that reading the value happens before incrementing the m_readPos
+    // otherwise the following can happen:
+    // 1. We increment m_readPos (but the value hasn't been read yet)
+    // 2. Another thread calls push(): we check if the queue is full => no
+    // 3. In push(), a data race can occur
+    // With memory_order_acq_rel,
+    // m_readPos must be increased after reading the popped value otherwise
+    // it is possible that the popped value is overwritten by push while being read
+    // (e.g. when the queue is full, i.e. m_readPos == m_writePos)
+    // Note that the following situation can still happen but is not a problem:
+    // 1. We read the value
+    // 2. Another thread calls push(): we check if the queue is full => yes
+    // 3. The queue was not really full as the value was already read
+    // Memory order acq_rel to avoid:
+    // - reading m_data to happen after the fence (rel)
+    // - incrementing m_readPos before the fence (acq)
+    std::atomic_thread_fence(std::memory_order::memory_order_acq_rel);
+    m_readPos.store(current_readPos + 1, std::memory_order_relaxed);
```

Comment:
I fail to see why this is required. Where is the race in the `push` method?

Are you confusing this with the SOFI?

Reply:
Initial situation: the queue is full
```
|--1--|--2--|
^
r=0,  w=2
```

We increment the read counter but haven't read the value yet (operation reordering)
```
|--1--|--2--|
^     ^
w=2  r=1
```

The pop thread is paused and push is called in another thread. The queue is not full so the push thread writes a new value:
```
|--3--|--2--|
      ^
   r=1, w=3
```

Now the pop thread resumes and reads the value at the previously stored index 0 => 3 (1 was never read)

Reply:
The idea here was to say:

1. there is no memory synchronization needed here (we are just reading) BUT we need to avoid operation reordering
2. We make the memory order `relaxed` to show that there is no memory synchronization
3. Since we still need to prevent operation reordering, we use an atomic fence

Reply:
Well, it seems Miri agrees with me https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=33665814b2deb0ef8b42cb7fe4255a71

Like already mentioned in another comment, adding either a release ordering on line 72 in the new link or a fence with acquire semantics on line 74 will fix the problem. Instead of the fences I would rather use the atomics as proposed in https://github.com/eclipse-iceoryx/iceoryx/pull/2168#discussion_r1474886638. I'm also fine with keeping the memory order as is and I create a separate PR to fix the issue.

Fences do not work like atomics and have subtle differences. Each time I work with fences I refresh my memory with these blog posts from Jeff Preshing.
- https://preshing.com/20130922/acquire-and-release-fences/
- https://preshing.com/20131125/acquire-and-release-fences-dont-work-the-way-youd-expect/
- https://preshing.com/20150402/you-can-do-any-kind-of-atomic-read-modify-write-operation/
- https://preshing.com/20170612/can-reordering-of-release-acquire-operations-introduce-deadlock/

I would need to dig in again in order to explain you exactly why this code does not work.

Reply:
Thanks for the links. I think there is a misunderstanding. Here is the behavior I observed:

| line 72: read_pos.load(order) | line 113: read_pos.store(order) | miri report |
|-------------------------------|---------------------------------|-------------|
| Relaxed                       | Release                         | race        |
| Relaxed                       | Relaxed + fence(Release)        | race        |
| Acquire                       | Relaxed                         | race        |
| Acquire                       | Release                         | no race     |
| Acquire                       | Relaxed + fence(Release)        | no race     |

So it looks like the memory fence is ok

Reply:
@albtam yes, that's what I meant. Keeping the current memory order and just adding an `Acquire` also fixes the race without the need of a fence. The fence would in this case just make the reasoning harder.

Reply:
Just to be sure to prevent a misuderstanding.
Current code
```cpp
push(...)
{
    m_readPos.load(relaxed); // via is_full
    // doing stuff
    m_writePos.store(release);
}
pop(...)
{
    m_readPos.load(acquire);
    m_writePos.load(acquire);
    // doing stuff
    m_readPos.store(release);
}
```

Your initial proposal
```cpp
push(...)
{
    m_readPos.load(relaxed); // via is_full
    // doing stuff
    m_writePos.store(release);
}
pop(...)
{
    m_readPos.load(relaxed);
    m_writePos.load(acquire);
    // doing stuff
    fence(acq_rel);
    m_readPos.store(relaxed);
}
```

Your updated proposal
```cpp
push(...)
{
    m_readPos.load(acquire);
    // doing stuff
    m_writePos.store(release);
}
pop(...)
{
    m_readPos.load(relaxed);
    m_writePos.load(acquire);
    // doing stuff
    fence(acq_rel);
    m_readPos.store(relaxed);
}
```

My proposal
```cpp
push(...)
{
    m_readPos.load(acquire);
    // doing stuff
    m_writePos.store(release);
}
pop(...)
{
    m_readPos.load(relaxed);
    m_writePos.load(acquire);
    // doing stuff
    m_readPos.store(release);
}
```

Reply:
> The fence would in this case just make the reasoning harder.

I got your point since fences might not work as you expect. My reasoning is the following: when you use a fence, you only want to avoid operation reordering (but you don't need to sync the memory). When you specify the memory order in the load/store operation, you want (either or both) to prevent operation reordering and sync the memory. That's why I believe using fences makes the intention clearer 

Note: I actually edited my comment but you were too fast to answer, my final proposal would be:

```cpp
push(...)
{
    m_readPos.load(relaxed);
    if is_full {
      return false;
   }
   // only at this point we need to prevent operation reordering
  fence(acquire);


pop(...)
{
    m_readPos.load(relaxed);
    m_writePos.load(acquire);
    // doing stuff
    fence(release)
    m_readPos.store(relaxed);
}
```

That being said, I would go with your proposal, since you're the maintainer ;)

Reply:
As for explaining why the original code was wrong, here is my interpretation:

even if you have a release/store (line 113), it doesn't do anything (i.e. prevent operation rendering) if there is no corresponding load/acquire. Do you agree?

Reply:
Here I would go with the save option and don't use fences as long as we do not have an extensive test suite and also run that suite on a weakly ordered CPU. As already mentioned, fences work differently than atomics and one can introduce subtle bugs. With the atomics it is simpler to reason about what happens and the fences can be added later once we are sure it does not break the code, not even on edge cases.

Reply:
... also, keep in mind that fences also **do synchronize** memory.

Reply:
@albtam I think I realize more and more where the confusion came from. You are right that in theory you do not need to synchronize the memory back to the push thread but you need to establish a happens before relationship. This can either be done by an atomic store with release ordering in one thread and an atomic load with acquire ordering in the second thread or with two fences and an atomic with relaxed memory order or a fence and an atomic which has the complementary memory order to the fence. Using just a fence on one thread does not help you since it has no counterpart to establish the happens before relationship. Keep in mind, the only reason you want the ordering is because there is another thread which can observe the ordering and therefore needs to be synced with.